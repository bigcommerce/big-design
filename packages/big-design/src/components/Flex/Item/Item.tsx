import React, { forwardRef } from 'react';

import { BoxProps } from '../../Box';
import { FlexedItemProps } from '../types';

import { StyledFlexItem } from './styled';

interface PrivateProps {
  readonly forwardedRef: React.Ref<HTMLDivElement>;
}

export type FlexItemProps = BoxProps & FlexedItemProps;

const RawFlexItem: React.FC<FlexItemProps & PrivateProps> = ({ as, forwardedRef, ...props }) => (
  <StyledFlexItem forwardedAs={as} ref={forwardedRef} {...props} />
);

export const FlexItem = forwardRef<HTMLDivElement, FlexItemProps>((props, ref) => (
  <RawFlexItem {...props} forwardedRef={ref} />
));
