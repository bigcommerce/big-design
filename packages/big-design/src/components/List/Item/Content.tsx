import { IconProps } from '@bigcommerce/big-design-icons';
import React, { cloneElement, isValidElement, memo, useCallback, useMemo } from 'react';

import { DropdownItem, DropdownLinkItem } from '../../Dropdown';
import { Flex, FlexItem } from '../../Flex';
import { SelectAction, SelectOption } from '../../Select';
import { Tooltip } from '../../Tooltip';
import { Small } from '../../Typography';

import { StyledLink } from './styled';

interface ContentProps {
  item: DropdownItem | DropdownLinkItem | SelectOption<any> | SelectAction;
  isHighlighted: boolean;
}

export const Content = memo(({ item, isHighlighted }: ContentProps) => {
  const iconColor = useMemo(() => {
    if (item.disabled) {
      return 'secondary40';
    }

    if (!isHighlighted || !('onActionClick' in item)) {
      return 'secondary60';
    }

    return 'actionType' in item && item.actionType ? 'danger50' : 'primary';
  }, [isHighlighted, item]);

  const renderIcon = useMemo(
    () =>
      isValidElement<IconProps>(item.icon) &&
      cloneElement(item.icon, {
        color: iconColor,
        size: 'large',
      }),
    [iconColor, item],
  );

  const descriptionColor = useCallback(
    (isDisabled: boolean | undefined) => (isDisabled ? 'secondary40' : 'secondary60'),
    [],
  );

  const wrapInLink = useCallback(
    (linkItem: DropdownLinkItem, content: React.ReactNode) => (
      <StyledLink href={linkItem.url} tabIndex={-1} target={linkItem.target}>
        {content}
      </StyledLink>
    ),
    [],
  );

  const wrapInTooltip = useCallback(
    (tooltip: string, tooltipTrigger: React.ReactElement) => (
      <Tooltip
        modifiers={[{ name: 'preventOverflow' }, { name: 'offset', options: { offset: [0, 20] } }]}
        placement="left"
        trigger={tooltipTrigger}
      >
        {tooltip}
      </Tooltip>
    ),
    [],
  );

  const getContent = useMemo(() => {
    const { content, disabled, description, icon } = item;

    const baseContent = (
      <Flex alignItems="center" flexDirection="row">
        {icon && (
          <FlexItem
            alignSelf={description ? 'flex-start' : undefined}
            paddingRight="xSmall"
            paddingTop={description ? 'xSmall' : undefined}
          >
            {renderIcon}
          </FlexItem>
        )}
        {description ? (
          <FlexItem paddingVertical="xSmall">
            {content}
            <Small color={descriptionColor(disabled)}>{description}</Small>
          </FlexItem>
        ) : (
          content
        )}
      </Flex>
    );

    const finalContent =
      'type' in item && item.type === 'link' && !disabled
        ? wrapInLink(item, baseContent)
        : baseContent;

    return disabled && 'tooltip' in item && item.tooltip
      ? wrapInTooltip(item.tooltip, finalContent)
      : finalContent;
  }, [descriptionColor, item, renderIcon, wrapInLink, wrapInTooltip]);

  return getContent;
});
