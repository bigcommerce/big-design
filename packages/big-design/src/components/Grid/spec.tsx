import { render, screen } from '@testing-library/react';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { GridItem } from './Item';

import { Grid } from './index';

test('render Grid', () => {
  const template = `
  "head head" 80px
  "nav  main" 1fr
  "nav  foot" 10%
  / 120px 1fr;
`;

  render(
    <Grid data-testid="grid" gridTemplate={template}>
      <GridItem gridArea="head">Header</GridItem>
      <GridItem gridArea="nav">Sidebar</GridItem>
      <GridItem gridArea="main">Content</GridItem>
      <GridItem gridArea="foot">Footer</GridItem>
    </Grid>,
  );

  expect(screen.getByTestId('grid')).toMatchSnapshot();
});

test('has display grid', () => {
  render(<Grid data-testid="grid" />);

  expect(screen.getByTestId('grid')).toHaveStyle('display: grid');
});

test('has gap properties', () => {
  const { rerender } = render(<Grid data-testid="grid" gridGap="3rem" />);

  expect(screen.getByTestId('grid')).toHaveStyle('gap: 3rem');

  rerender(<Grid data-testid="grid" gridColumnGap="1rem" gridRowGap="2rem" />);

  expect(screen.getByTestId('grid')).toHaveStyle({ 'row-gap': '2rem', 'column-gap': '1rem' });
});

test('Grid forwards styles', () => {
  render(<Grid className="test" data-testid="grid" style={{ background: 'red' }} />);

  expect(screen.getByTestId('grid')).toHaveClass('test');
  expect(screen.getByTestId('grid')).toHaveStyle('background: red');
});

test('Grid item forwards styles', () => {
  render(<GridItem className="test" data-testid="grid-item" style={{ background: 'red' }} />);

  expect(screen.getByTestId('grid-item')).toHaveClass('test');
  expect(screen.getByTestId('grid-item')).toHaveStyle('background: red');
});

test('rendering as another element retains inherited props and styles', () => {
  render(<Grid as="section" data-testid="grid" margin="medium" />);

  const grid = screen.getByTestId('grid');

  expect(grid.tagName).toBe('SECTION');
  expect(grid).toHaveStyle(`margin: 1rem`);
});

test('grid forwards ref', () => {
  const ref = createRef<HTMLDivElement>();

  render(<Grid data-testid="grid" ref={ref} />);

  const grid = screen.getByTestId('grid');

  expect(grid).toBe(ref.current);
});
