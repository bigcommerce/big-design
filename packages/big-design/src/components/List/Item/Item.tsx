import { CheckIcon } from '@bigcommerce/big-design-icons';
import { UseComboboxPropGetters, UseSelectPropGetters } from 'downshift';
import React, { ComponentPropsWithoutRef, forwardRef, memo, Ref } from 'react';

import { typedMemo } from '../../../utils';
import { Checkbox } from '../../Checkbox';
import { DropdownItem, DropdownLinkItem } from '../../Dropdown';
import { SelectAction, SelectOption } from '../../Select';

import { Content } from './Content';
import { StyledListItem } from './styled';

export interface ListItemProps<T> extends ComponentPropsWithoutRef<'li'> {
  readonly actionType?: 'normal' | 'destructive';
  readonly autoWidth?: boolean;
  readonly disabled?: boolean;
  readonly index: number;
  readonly isAction?: boolean;
  readonly isChecked?: boolean;
  readonly isHighlighted: boolean;
  readonly isIndeterminate?: boolean;
  readonly isSelected?: boolean;
  readonly item: DropdownItem | DropdownLinkItem | SelectOption<T> | SelectAction;
  readonly getItemProps:
    | UseSelectPropGetters<any>['getItemProps']
    | UseComboboxPropGetters<any>['getItemProps'];
  addItem?(item: SelectOption<T>): void;
  removeItem?(item: SelectOption<T>): void;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLLIElement>;
}

const StyleableListItem = typedMemo(
  <T,>({
    actionType = 'normal' as const,
    autoWidth = false,
    forwardedRef,
    index,
    isAction = false,
    isChecked = false,
    isHighlighted,
    isSelected = false,
    isIndeterminate = false,
    item,
    getItemProps,
    addItem,
    removeItem,
    ...props
  }: ListItemProps<T> & PrivateProps): ReturnType<React.FC<ListItemProps<T> & PrivateProps>> =>
    removeItem && addItem ? (
      <StyledListItem
        {...getItemProps({
          ...props,
          index,
          item,
          onClick: () => {
            if (item.disabled) {
              return;
            }

            const hasValue = (
              item: DropdownItem | DropdownLinkItem | SelectOption<T> | SelectAction,
            ): item is SelectOption<T> => 'value' in item;

            if (hasValue(item)) {
              isChecked ? removeItem(item) : addItem(item);
            }
          },
          ref: forwardedRef,
        })}
        actionType={actionType}
        autoWidth={autoWidth}
        isAction={isAction}
        isHighlighted={isHighlighted}
      >
        <Checkbox
          checked={isChecked}
          description={item.description}
          disabled={item.disabled}
          isIndeterminate={isIndeterminate}
          label={item.content}
          onChange={() => null}
          onClick={(event) => event.preventDefault()}
          tabIndex={-1}
        />
      </StyledListItem>
    ) : (
      <StyledListItem
        {...getItemProps({
          ...props,
          index,
          item,
          ref: forwardedRef,
        })}
        actionType={actionType}
        autoWidth={autoWidth}
        disabled={item.disabled}
        isAction={isAction}
        isHighlighted={isHighlighted}
        isSelected={isSelected}
      >
        <Content isHighlighted={isHighlighted} item={item} />
        {isSelected && <CheckIcon color="primary" size="large" />}
      </StyledListItem>
    ),
);

export const ListItem = memo(
  forwardRef<HTMLLIElement, ListItemProps<unknown>>((props, ref) => (
    <StyleableListItem {...props} forwardedRef={ref} />
  )),
);
