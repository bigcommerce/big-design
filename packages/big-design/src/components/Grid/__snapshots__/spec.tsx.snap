// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render Grid 1`] = `
.c0 {
  box-sizing: border-box;
}

.c2 {
  grid-area: head;
}

.c3 {
  grid-area: nav;
}

.c4 {
  grid-area: main;
}

.c5 {
  grid-area: foot;
}

.c1 {
  gap: 1rem;
  grid-template: "head head" 80px "nav  main" 1fr "nav  foot" 10% / 120px 1fr;
  display: grid;
}

<div
  class="c0 c1"
  data-testid="grid"
>
  <div
    class="c0 c2"
  >
    Header
  </div>
  <div
    class="c0 c3"
  >
    Sidebar
  </div>
  <div
    class="c0 c4"
  >
    Content
  </div>
  <div
    class="c0 c5"
  >
    Footer
  </div>
</div>
`;
